# Sum the number of comments, kudos , hits, and the number of works

For each character, the number of of stories they appear in, and the total number of comments, kudos and hits those stories attracted.

Similarly for each slash pairing.

## Load data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt

import ds9
df = ds9.df()

## Drop the unneeded columns

If you want to look into other columns, then add them here.

In [2]:
df = df[['charactersclean', 'relationshipspairslash', 'comments', 'hits', 'kudos', 'words']]

# Program definitions

### allrows()

Find the rows which match the column's content. Return them in popularity order.

In [3]:
def allrows(df, match):
    # Explode the dataframe by column of interest into new_df
    new_df = pd.DataFrame()
    for index, matches in zip(df.index, df[match]):
        for i in matches:
            row = df[df.index == index].copy()
            row[match] = i
            new_df = pd.concat([new_df, row], ignore_index=True)
    # Get the characters and their frequency
    f = new_df[match].value_counts()
    f = f.reset_index(name = 'n')
    return f['index']

### summarisecharacter()

Go through the list of characters from the step above. For each character: find all the rows of the dataframe where they are in some `column` (like `charactersclean`); count the number of those rows; sum the values of the columns `comments`, `kudos`, `hits` and `words`.

In [4]:
def summarisecharacter(df, character, column):
    character_df = pd.DataFrame()
    for index, characters in zip(df.index, df[column]):
        if character in characters:
            # Match found, copy the whole row to the bottom of the new dataframe
            row = df[df.index == index]
            character_df = pd.concat([character_df, row])
    # Easiest way to create a dataframe is to populate a dictionary of lists.
    # Since this is a one-row dataframe, the lists in [ ] have one item.
    d = dict()
    d[column] = [ character ]
    d['works'] = [ character_df.shape[0] ]
    d['comments'] = [ character_df['comments'].sum() ]
    d['kudos'] = [ character_df['kudos'].sum() ]
    d['hits'] = [ character_df['hits'].sum() ]
    d['words'] = [ character_df['words'].sum() ]
    return pd.DataFrame.from_dict(d)

### summariseallcharacters()

Return a dataframe of all the characters found in the column with a summary of the works statistics for that character.

In [5]:
def summariseallcharacters(df, column):
    ac = allrows(df, column)
    sum_df = pd.DataFrame()
    for character in ac:
        summary = summarisecharacter(df, character, column)
        sum_df = pd.concat([sum_df, summary], ignore_index=True)
    sum_df.set_index(column, inplace=True)
    return sum_df

## Comments, hits and kudos for characters

That is, for `charactersclean`.

In [6]:
sum_df = summariseallcharacters(df, 'charactersclean')
sum_df.to_csv('sum-comments-hits-kudos-charactersclean.csv')
pd.options.display.max_rows = 999
sum_df

works  comments   kudos     hits     words
charactersclean                                                               
Julian Bashir                        4938    110911  633161  6343689  29428092
Elim Garak                           4521    106712  606841  5971556  24994180
Kira Nerys                           1945     38681  142215  1791858  16928641
non-cast                             1922     49296  147675  1839595  21799734
Jadzia Dax                           1631     35007  163586  1743817  12520276
Odo                                  1380     28614  127886  1444062  11561861
Benjamin Sisko                       1190     27159  107504  1233615  12703665
Quark                                1122     24492  108962  1091406   9433072
Miles O'Brien                        1051     35308  138997  1452537  10809417
Ezri Dax                              577     12505   34412   479394   5297759
Skrain Dukat                          544     10131   34805   507387   5955291
Worf                                  438     10551   35947   411843   4874295
Jake Sisko                            384      7415   29832   324676   3846649
Weyoun                                369      3830   16622   189508   2235686
Keiko O'Brien                         365     15483   44381   459623   4305369
Corat Damar                           323      4613   16014   196563   2637656
Tora Ziyal                            289      6764   21937   276362   3031499
Nog                                   283      4959   20609   213615   2862848
Kelas Parmak                          224     10336   17849   254639   2465464
Rom                                   203      4864   14346   148840   2300193
Enabran Tain                          186      7574   21931   232482   1861477
Leeta                                 169      4538   14734   164663   1818890
Molly O'Brien                         147      5941   17048   149749   1436653
Kasidy Yates                          121      2463    6937    91487    867974
Martok                                 91      4116   10018   122986   1231676
Lenara Kahn                            87       850    2859    46102    423522
Morn                                   85      1852    7125    73317    613010
Mila Garak                             73      3437    8016    74290    609202
Lwaxana Troi                           73      2778    7515    83611    803252
Amsha Bashir                           69      2434    9019    93746   1239655
Winn Adami                             68       516    1333    26828    984078
Richard Bashir                         67      1991    8273    81731   1110682
Shakaar Edon                           63       991    1935    32255   1159692
Vic Fontaine                           62      1069    4475    55037    766861
Kirayoshi O'Brien                      54      1000    4853    49666    363787
Female Changeling                      52       626    2259    32536    499331
Kukalaka                               51      2923    6914    68678    577842
Bareil Antos                           50       577    1446    18267    741079
Dax                                    48       334    1144    14492    329796
Q                                      48       760    1609    29780    761050
Tora Naprem                            47       726    1745    25388    428883
Jabara                                 47      2479    6274    67209    492165
Keevan                                 43       383    1171    19352    394972
Ishka                                  35       651    1804    20521    275559
Curzon Dax                             33       354    1809    17909    138067
Brunt                                  32       465    1488    12244    114727
Luther Sloan                           32       703    2010    31604    551386
Alexander Rozhenko                     30       851    1504    26886    800421
Sarina Douglas                         29       796    1789    26961    292600
Kira Me

### Interpreting these results

There are 110911 comments in works in which Julian Bashir is tagged as a character (the tagging done by the author of the work).

Works tagged with the character Julian Bashir are more downloaded than works tagged with the character Elim Garak.

Note that 'non-cast' refers to the sum of all non-cast characters, not just one examplar character.

## Comments, hits and kudos for slash pairings

That is, for `relationshipspairslash`.

In [7]:
sum_df = summariseallcharacters(df, 'relationshipspairslash')
sum_df.to_csv('sum-comments-hits-kudos-relationshipspairslash.csv')
pd.options.display.max_rows = 999
sum_df

works  comments   kudos  \
relationshipspairslash                                                        
Elim Garak/Julian Bashir                             3812     96437  600400   
Jadzia Dax/Kira Nerys                                 394      4724   25445   
Odo/Quark                                             372      6319   38204   
non-cast/non-cast                                     304      7002   17033   
Kira Nerys/Odo                                        244      2509   14683   
Jadzia Dax/Worf                                       144      2776   10460   
Julian Bashir/non-cast                                143      3823   10257   
Keiko O'Brien/Miles O'Brien                           125      4256   13185   
Corat Damar/Weyoun                                    124      1527    6412   
Elim Garak/Kelas Parmak                               118      4094    7332   
Ezri Dax/Julian Bashir                                112      4624    6457   
Julian Bashir/Miles O'Brien                            98      1313    6576   
Elim Garak/non-cast                                    88      3323    7573   
Jake Sisko/Nog                                         84      1237    6283   
Ezri Dax/Kira Nerys                                    72       587    2617   
Jadzia Dax/Lenara Kahn                                 67       511    2553   
Benjamin Sisko/Kasidy Yates                            61       890    2206   
Kira Nerys/Skrain Dukat                                59       982    4132   
Skrain Dukat/non-cast                                  56      1499    2614   
Elim Garak/Julian Bashir/Kelas Parmak                  47      1849    3418   
Jadzia Dax/Julian Bashir                               42       322    2583   
Benjamin Sisko/Kira Nerys                              42       308     654   
Elim Garak/Skrain Dukat                                39       412    2385   
Elim Garak/Tora Ziyal                                  36       603    2148   
Corat Damar/Kira Nerys                                 35       614    1668   
Kira Nerys/non-cast                                    34       579    1421   
Elim Garak/Odo                                         33       125    1038   
Keiko O'Brien/Kira Nerys/Miles O'Brien                 33       782    4013   
Weyoun/non-cast                                        33       181     872   
Skrain Dukat/Tora Naprem                               33       188     717   
Benjamin Sisko/Julian Bashir                           33       298     890   
Bareil Antos/Kira Nerys                                32       302     465   
Jadzia Dax/non-cast                                    32       347    1024   
Skrain Dukat/Weyoun                                    32       463    1613   
Julian Bashir/Leeta                                    29      1410    4107   
Jake Sisko/Tora Ziyal                                  29       565    2151   
Leeta/Rom                                              28       448    1799   
Keiko O'Brien/Kira Nerys                               27       302    1511   
Julian Bashir/Kira Nerys                               27      1053    1759   
Julian Bashir/Kelas Parmak                             26      1481    2056   
Corat Damar/non-cast                                   23       592     956   
Ezri Dax/Lenara Kahn                                   23       391     795   
Corat Damar/Skrain Dukat                               22       483    1449   
Kasidy Yates/Kira Nerys                                21        71     425   
Benjamin Sisko/Skrain Dukat                            21       266    1385   
Kira Nerys/Kira Nerys                                  21       141     696   
Julian Bashir/Skrain Dukat                             20       962    2852   
Brunt/Quark                                            20       336    1213   
Quark/non-cast                                         19       216     658   
Kira Nerys/Shakaar Edon